glean-core/tests/event.rs

```diff
@@ -0,0 +1,245 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+mod common;
+use crate::common::*;
+
+use std::collections::HashMap;
+use std::fs;
+
+use glean_core::metrics::*;
+use glean_core::{CommonMetricData, Glean, Lifetime};
+
+#[test]
+fn record_properly_records_without_optional_arguments() {
+    let (_t, tmpname) = tempdir();
+    let store_names = vec!["store1".into(), "store2".into()];
+
+    let glean = Glean::new(&tmpname, GLOBAL_APPLICATION_ID, true).unwrap();
+
+    let metric = EventMetric::new(
+        CommonMetricData {
+            name: "test_event_no_optional".into(),
+            category: "telemetry".into(),
+            send_in_pings: store_names.clone(),
+            disabled: false,
+            lifetime: Lifetime::Ping,
+        },
+        vec![],
+    );
+
+    metric.record(&glean, 1000, None);
+
+    for store_name in store_names {
+        let events = metric.test_get_value(&glean, &store_name).unwrap();
+        assert_eq!(1, events.len());
+        assert_eq!("telemetry", events[0].category);
+        assert_eq!("test_event_no_optional", events[0].name);
+        assert!(events[0].extra.is_none());
+    }
+}
+
+#[test]
+fn record_properly_records_with_optional_arguments() {
+    let (_t, tmpname) = tempdir();
+
+    let store_names = vec!["store1".into(), "store2".into()];
+
+    let glean = Glean::new(&tmpname, GLOBAL_APPLICATION_ID, true).unwrap();
+
+    let metric = EventMetric::new(
+        CommonMetricData {
+            name: "test_event_no_optional".into(),
+            category: "telemetry".into(),
+            send_in_pings: store_names.clone(),
+            disabled: false,
+            lifetime: Lifetime::Ping,
+        },
+        vec!["key1".into(), "key2".into()],
+    );
+
+    let extra: HashMap<i32, String> = [(0, "value1".into()), (1, "value2".into())]
+        .iter()
+        .cloned()
+        .collect();
+
+    metric.record(&glean, 1000, Some(extra));
```

Comment:
Seeing this now, there's a trick in Rust to make both `metric.record(&g, 1000, None)` and `metric.record(&g, 1000, extra)` work.

```rust
fn record<M: Into<Option<HashMap<...>>>(glean: &Glean, timestamp: u64, extra: M) {
  let extra = extra.into()
  ...
}
```

Just a suggestion, maybe we don't need this here, it's not made to be the nice user-facing API (yet)

Reply:
Cool trick.  Might as well do this, I guess...